### DISCUSSION TITLE
Generating a Tree-Editor model schema from an eCore file

### DISCUSSION THREAD

**Discussion by Reeze-Moodaley** [eclipse-emfcloud/emfcloud#142] [Theia Tree Editor] [answered]
After doing some digging into the schemas used by the tree editor and our eCore file, we would like to know whether it would be possible for us to generate our model schema from our eCore file as this would save us the time of having to manually create schemas for a model as massive as ours and we would not have to repeat information that is already present within our eCore ( Such as what nodes may be children of other nodes).

Thank you in advance!

**Comment by vhemery** [comment, marked as answer]
You should be able to customize everything that is generated with org.eclipse.emfcloud.modelserver.jsonschema.JsonSchemaConverter and org.eclipse.emfcloud.modelserver.emf.common.SchemaController. (both UI schema and model schema)
There are default implementations binded by your Guice module, and you can override them.

**Reply by agonzalezgaliana** [reply]
What I understand from the question is how to handle the "tree model" (i.e. coffee-model.ts in the Coffee Editor example) when your EMF model (*.ecore) is big and complex.

From my point of view (based on doing some reverse engineering of the Coffee Editor source code), what is defined in the coffee-model.ts is a kind of "duplication" of what is already defined int the coffe.ecore file. On one side we have the Type definition:

`
export const AutomaticTask = 'http://www.eclipse.org/emfcloud/coffee/model#//AutomaticTask';
...
`

And on the other side we define the "creatable" children map of each one of the types pointing for each type of node what children can be created.

In the case of big EMF models, the manual creation of the creatable children map could be tedious (also for maintenance, in case that the EMF model changes, to rework a *.ts file on the Theia extension).

**Reply by vhemery** [reply]
I interpreted the question as the generation of JsonForms schemas.
Indeed, maybe the intention was generating the model Typescript classes corresponding to the ecore model.

For that second option, this would be a new feature request which would IMHO be very valuable. It is quite straightforward to implement in xtend and could be plugged in the EcoreTools tooling.
Moreover, this would lead to a normalization of the usages of the EMF Cloud Model Server, and other helpfull tooling could be added on top.
Such generation does not exist yet to my knowledge.

**Reply by eneufeld** [reply]
So for the coffee editor the schemas are [here](https://github.com/eclipse-emfcloud/coffee-editor/blob/master/web/coffee-editor-extension/src/browser/coffee-tree/coffee-schemas.ts).
Yes the schema is basically a duplicate of the ecore.
Unfortunately we need this map at the moment: https://github.com/eclipse-emfcloud/coffee-editor/blob/master/web/coffee-editor-extension/src/browser/coffee-tree/coffee-model.ts#L53 

EMFForms has a feature to export the viewmodels and corresponding ecores to schema and ui schema. You could use this to get a head start. See this blog post: https://eclipsesource.com/blogs/2017/07/26/emf-forms-1-13-0-feature-export-to-json-forms/
EMFForms is currently in version 1.26 but the feature is the same.

**Reply by agonzalezgaliana** [reply]
> 
> ...
Unfortunately we need this map at the moment: https://github.com/eclipse-emfcloud/coffee-editor/blob/master/web/coffee-editor-extension/src/browser/coffee-tree/coffee-model.ts#L53
> ...

This map can be auto generated once we get the JSON schema from the Model Server. I've already implemented a way to generate it, however I'm still not able to use it. It seems that the map is queried when the Tree Editor contributor is initialized (see [here](https://github.com/eclipse-emfcloud/theia-tree-editor/blob/master/theia-tree-editor/src/browser/util.ts#L81)) and never refreshed. In my case I'm parsing the JSON schema from the widget initialization before loading the model. Would it make sense to query the JSON schema and parse it before the "creatableChildrenMapping" is queried from the "generateAddCommandDescriptions" method?

**Reply by lucas-koehler** [reply]
Hi @agonzalezgaliana , the command contributions are executed during Theia's startup (as it is usually done, see [Theia docs](https://theia-ide.org/docs/authoring_extensions)). Consequently, fetching the JSON Schema before this is likely not possible with this setup. You could get your JSONSchema and generate the `creatableChildrenMap`  in the model service the first time its method `getChildrenMapping` is called (which should be by `generateAddCommandDescriptions`).

The remaining problem is that doing this could slow down Theia's startup time.

**Reply by agonzalezgaliana** [reply]
Hi @lucas-koehler, I see ... That's why in the Coffee Editor example "everything" in the model service is somehow hard coded. Perhaps "could" make sense, since a JSON schema is not intended to be changed that often, to hard code the map it in a TS file. However I find it "cleaner" to build that map from the JSON schema retrieved from the Model Server.

Anyways thanks for your answer!

**Reply by lucas-koehler** [reply]
@agonzalezgaliana I agree that for bigger models or ones that change more often the generation is cleaner. If you fetch it from a backend, maybe consider pre-generating the child mapping?
Another thing you can try in addition is to register the commands asynchronously to avoid slowing down start up. [Here](https://spectrum.chat/theia/dev/menu-contributions-depending-on-the-backend~86e6e098-1413-4b31-8a20-eef54ca6ff11) this was down with menu contributions. Might also work with the commands by making the `registerCommands` method `async`.

**Reply by agonzalezgaliana** [reply]
@lucas-koehler Thanks for the provided tips. I'll take a look on the async commands. With regards to the pre-generation of the child mapping, I'm not sure if I get what you mean. Do you mean to fetch the child mapping from a server instead of fetching the JSON schema + generating the map in the client side? May this be possible in the Model Server? How possible it would be to enhance the V1/V2 REST APIs?

**Reply by lucas-koehler** [reply]
@agonzalezgaliana Yes, that is what I meant. However, this would not solve your problem in general but just increase performance on the client side. Unfortunately, I personally don't know how to extend the Model Server REST APIs. But if you want to do that I suggest opening a separate question for that :)
